### ISSUE TITLE
Possible re-design of event filters

### ISSUE THREAD

**Issue by HDembinski** [impy-project/impy#21]
From the discussion of #20 
> For the filters, there is maybe no deep reason. I personally think that numpy boolean selections or np.where-s are a sufficient filtering framework. But preselecting charged and stable particles is so common that requiring always an additional expression `event.status == 1` in your selections is quite annoying every time. I think what is needed are just these two "pre-filters" (for final state and charge) and the rest is up to the user to select stuff, e.g. something like `np.where((event.p_ids == 211) & (np.abs(event.eta) < 2.5))` or way more complex expressions.

Ok, +1 for adding a short for common filters, but ideally, we should provide a simple and familiar system for filtering for the user and then make applying the standard filters an easy-to-use special case. It is good to write the API so that filtering is done always in the same way, even for common filtering operations.

Think about it from the user side. If you provide a special method that does some filtering, but the function does not show how to do the filtering in general, then the user has to learn two things: your special method and the general way. It is more educational to only show the general way and make that simple to understand, so there is "one obvious way" to do it (see Zen of Python, which is excellent design advice that should be taken very seriously, despite the light-hearted tone). This way, people automatically learn how to do powerful filtering from the commonly applied simple filtering.

Here is some example how such filtering could look like. We want to mask all arrays at once when we apply a filter, so we need to be able to apply masks directly on the event object, like so
```py
filtered_event = event[mask]
```
This would require implementing `__getitem__` on the event object, which then forwards the mask to its member arrays and constructs a new event object with the filtered contents.

However, this is a bit inelegant and inefficient. It would be more elegant and efficient if the event object itself maps to the FORTRAN common memory block. event should be a subclass of numpy.ndarray or numpy.ma.masked_array. Subclassing ndarray is not easy, but I have done it before and it is well supported by numpy (even if the process is a bit cumbersome).

If event itself is a numpy array view of the underlying FORTRAN memory block, then masking particles becomes a trivial operation that we do not even have to implement. For convenience, event could still have properties like `event.px` that return a subview into all px values etc. So the interface visible to the user can remain the same.

Some words on masked arrays. Numpy's masked arrays could be a great match for this application. Masked arrays allow one to filter out particles without physically changing the value arrays. The masks act non-destructively, so "clearing filters" remains possible. All that happens without copying any data either.

**Comment by afedynitch** [comment]
Ohh... Now I understand what you mean. Right now the basic atrributes (px, py..., E, m, p_ids, etc.) map directly to the fortran common blocks as views. The current filter scheme is indeed quite horrible in the sense that it destructively applies views on all variables if filters are used. I think it is not a big deal to change the architecture to the different style of filtering.

For impy this shouldn't be something like this in `MCEvent`:
```
def __get_item__(self, key):
     self._mask = key
     return self
```
Then each user-accessible variable becomes a property, e.g.:
```
@property
def px(self):
    return self.px[self._mask]
    # Or in case of masked arrays
    self.px.mask = self._mask
    return self.px
```
For all derived variable, such as $p_T^2$, there is no change because they already are properties and will see the mask immediately.
```
@property
def pt2(self):
    """Transverse momentum squared in (GeV/c)**2"""
    return self.px**2 + self.py**2
```

On the user side one needs to construct masks like that??:
```
for event in generator(20):
    stable = np.where(event.status == 1)
    pions = np.where(np.abs(event[stable].p_ids) == 211)
    all_pion_pt = event[pions].pt 
```

I'm not super familiar with masked arrays and use them occasionally. Although, I'm not against using them in general, my experience is that sometimes weird things happen that need specific workarounds. Also, not sure about the speed since I never benchmarked masked arrays in more detail.

**Comment by HDembinski** [comment]
Yes, that is more or less what I had in mind, but I think for this use case, we could simplify the code we have to main by using numpy.ma.masked_array. I have to experiment a bit with that. I am not sure whether you can create a masked array as a view into an external memory buffer. I don't see why not, though.

**Comment by HDembinski** [comment]
Then you would do
```py
for event in generator(20):
    is_stable = np.where(event.status == 1)
    is_pion = np.abs(event.p_ids) == 211
    # filter out stable pions
    event.mask = is_stable & is_pion
    # do something with event.px
    # clear filter, no large memory buffers are created or destroyed at any point
    event.mask = True # numpy broadcasting
```

**Comment by HDembinski** [comment]
For common filtering operations, the library can provide small helper functions:
```
def is_stable(event):
    return event.status == 1

def is_pion(event):
    return np.abs(event.p_ids) == 211
```
etc

**Comment by HDembinski** [comment]
What would really help is to add some status enum which decodes the number 1 for status into something readable and using particle to replace raw pdg ids.
```py
from particle import literals as lp

class Status(enum.Enum):
    final = 1
    # other status flags

for event in generator(20):
    event.mask = (np.abs(event.p_ids) == lp.pi_plus.pdgid) & (event.status == Status.final)
    # do something with event.px
    # clear filter, no large memory buffers are created or destroyed at any point
    event.mask = True # numpy broadcasting

**Comment by HDembinski** [comment]
@afedynitch @jncots While working fixing the HepMCWriter, I found bugs in the way parents and children are reported. While looking into this, I got overwhelmed by the complexity and the repeated code in the *Event classes. I am currently working on changes to the implementations of these classes. If possible, please do not touch these classes for now until I am done to avoid merge conflicts.

**Comment by afedynitch** [comment]
> What would really help is to add some status enum which decodes the number 1 for status into something readable and using particle to replace raw pdg ids.

The status is a mixed bag. For simple generators, like SIBYLL there is not much meaning of status. For a fully exposed stack of DPMJET or PYTHIA there are many many meanings of status. Even I don’t know all possible status codes for DPMJET. One may have multiple attributes for status, the true status and something less complex but it’s impossible to encode all status meanings for all generators consistently, and also would make it difficult to work with the manuals for the codes (like in case of PYTHIA where manuals actually exist).

Same for particles. It’s important to expose the PDG id for the standard (medium expert) user. In some generators the particle ID may encode something outside the standard (like for remnants in DPMJET). `Particle` comes with performance impact. So it could be something lazy evaluated as alternative to the standard pdg ids.

**Comment by afedynitch** [comment]
> While working fixing the HepMCWriter, I found bugs in the way parents and children are reported.

So far nobody worked with parents and children seriously, so it doesn’t surprise that you’re finding bugs. I usually need the history for generator development but usually look at tabulated printouts of events instead of using the variables in Event classes. Thanks for picking up this effort.

**Comment by HDembinski** [comment]
> The status is a mixed bag. For simple generators, like SIBYLL there is not much meaning of status. For a fully exposed stack of DPMJET or PYTHIA there are many many meanings of status. Even I don’t know all possible status codes for DPMJET. One may have multiple attributes for status, the true status and something less complex but it’s impossible to encode all status meanings for all generators consistently, and also would make it difficult to work with the manuals for the codes (like in case of PYTHIA where manuals actually exist).

An enum does not have to have fields for all possible status flags, but at least for the common ones. A minimal version is
```py
class Status(enum.Enum):
    final = 1
```

> Same for particles. It’s important to expose the PDG id for the standard (medium expert) user. In some generators the particle ID may encode something outside the standard (like for remnants in DPMJET). Particle comes with performance impact. So it could be something lazy evaluated as alternative to the standard pdg ids.

What I had in mind here is already implemented in https://github.com/impy-project/impy/pull/43, the performance impact is negligible. Another use case for `particle` is code like this:
```py
from particle import literals as lp

for event in generator(20):
    stable = np.where(event.status == 1)
    pions = np.where(np.abs(event[stable].pid) == lp.pi_plus.pdgid)
    all_pion_pt = event[pions].pt 
```
where the impact is negligible. Using a name is better than asking the user to remember cryptic numbers.

**Comment by HDembinski** [comment]
Fixed by #52
